### DISCUSSION TITLE
Helix DB planning

### DISCUSSION THREAD

**Discussion by xav-db** [HelixDB/helix-db#6] [Ideas]
This discussion is focussed on the planning of the project. How we will actually implement everything and the approaches we should take. 

The initial technical plan is to write everything in Rust. It is a strongly typed and powerful language which I believe to be a good choice for a project such as this.

**Comment by xav-db** [comment]
Initial ideas for basic query

```rust
use Schema::{User}                                          // import the schema

QUERY GetFollowers(userID: String) =>                       // declare new query
    GET Followers <- User(userID)::In::Follows                   // get all the followers of a user returning the user
        WHERE User::Status::Active                          // filter out followers who are not active
    LIMIT 50                                                // limit the number of followers to 50
    RETURN User::{Username, FollowerCount, FollowingCount}  // return the username, follower count, and following count of the followers
```
One thing I am unsure about is whether to start the query with something similar to gremlin like doing G::User or V::User to some other way to explicitly show the start of a traversal. 

Idea for being able to return group together results to be returned as a sub field (thing project by in gremlin)
Node Values are automatically unwrapped unless explicitly told otherwise, i.e. the user automatically unwraps into it's full object
Custom projections can be done in a separate function then applied here.
```rust
use Schema::{Restaurant};

QUERY GetRestaurantAndReviews(restaurantID: String) =>
    GET Res <- Restaurant                               // same as g.V().as('res')
    ADD_FIELD Res {                                     // Access fetched node      
        Reviews: [                                      // Define name for new field, square brackets indicate list, curly braces indicate single object
            Rating <- InE::Review::Rating               // Get the rating from the incoming edge of type Review - Same as Res::InE::Review::Rating
            User <- OutV::User                          // Get the user from the outgoing edge of type User - Same as Rating::OutV::User
        ]
    }
    RETURN Res
```

Equivalent gremlin query:
```rust
g.V(restaurantID).as('Restaurant')
    .inE('Review')
    .values('Rating').as('Rating')
    .outV().as('User')
    .project('Restaurant')
    .by(
        project('Reviews'/* SOME OTHER FIELDS IN HERE FOR REST OF RESTAURANT*/ ) 
        .by(Project('Rating', 'User')
            .by(select('Rating'))
            .by(
                project('Username', 'Email')
                .by(select('User').values('Username'))
                .by(select('User').values('Email'))
            )
        )
        // MORE PROJECTIONS FOR REST OF RESTAURANT
    )
    .toList()
```

**Reply by willbowden** [reply]
> ```GET User <- User(userID)::In::Follows```

Does this mean "Get all the `User`s following the user with `id=userID`"?  If so, is it wise to have the leftmost `User` be both the defining type of node to be retrieved and the name to access the resulting set of users? Or is the leftmost `User` just an identifier for the result of the `GET` operation. In that case, would it possibly be more readable to say: 
`Followers <- GET User(userID)::In::Follows`, or even use an `=` sign?

Also, it seems like the `::` operator is being used to access properties on an object. For example `User::Status::Active`. Is it then wise to have `In` and `OutV` be accessed in the same way? For example, should `User::In::Follows` be something like `User::In(Follows)`, or some other operator/symbol to denote traversal operations rather than property access?

Finally a minor point, but does the edge type `Follows` need to be accessed via a string identifier or does it exist as a concrete part of the schema?

Perhaps having a concrete example of what `Schema::{User}` would look like would help me understand more clearly.

**Reply by xav-db** [reply]
You're right!
> `GET User <- User(userID)::In::Follows`.

Is meant to be 
`GET Followers <- User(userID)::In::Follows`

The reason I did the assignment after the `GET` is so you could do something like the following

```rust
QUERY GetAllUsersAndFollowers =>
    GET User <- User
        Followers <- In::Follows
    RETURN User::{Username, FollowerCount}, Followers::{Username, FollowerCount}
```
Which would return 
```text
[
    [
        User: {
            Username: "John",
            FollowerCount: 10
        },
        Followers: [
            {
                Username: "Jane",
                FollowerCount: 20
            },
            ...
        ]
    ],
    ...
```
(Note: the return is not explicitly json, this is more of a generalisation of what the data might look like, which in this case is a bit like a tuple)

The `<-` operator simply looked nice to my eyes in the code which is why I used it. 

In terms of the schema, it might look like:
```rust
// Defines node type
User {
    Username: String,
    FollowerCount: U32,
}

// Defines outgoing edges for user 
User::Out {
     // Defines outgoing edge called "Follows" that no attributes that points to a User node.
     Follows<User> {}
}

// Could also explicitly define incoming nodes but the fact the outgoing edge points to the same type this would be inferred
```
> Finally a minor point, but does the edge type Follows need to be accessed via a string identifier or does it exist as a concrete part of the schema?
> Perhaps having a concrete example of what Schema::{User} would look like would help me understand more clearly.

Because of this schema definition structure, this is why I was thinking `User::In::Follows` as being like a property access.

**Reply by willbowden** [reply]
Okay, sounds good. In terms of the actual graph engine executing this query, how do you suppose it will run the example we're discussing here?

For example, in
```rust
GET User <- User
    Followers <- In::Follows
```

How would the engine know to traverse the `In` edges from the `User`s? Would/should that be explicitly mentioned in the query definition?
I'm just thinking if, for example, you wanted to make two traversals in one query that started at different points to aggregate data, rather than what seems to be happening above: assuming once the `User <- User` step is done that we are operating within the `User` space and can just write `In::Follows`, how would you specify a new starting point?

**Reply by xav-db** [reply]
Good point, this is why I was unsure in my original post about having a starting point like `g.V().....` in gremlin. i.e:
```rust
GET User <- V::User
        Followers <- User::In::Follows
```

In terms of having a different start point, that sounds like the anonymous traversals in gremlin. So you could either do a separate traversal with a GET (i.e. start again) or you could label it as anonymous i.e: 
```rust
GET User <- V::User
GET Post <- V::Posts
```
would return `{ [ User: {...}, User: {...}, ...], [ Post: {...}, Post: {...} ] }
OR
```rust 
GET User <- V::User(userID)
        WHERE (
                User::Has(COUNT ?::OutE)   // OR could be User::Has(?::OutE::Count) OR User::Has(Count(?::OutE))
        )
```

I've used the `?` to signify the anonymous operator but this could be anything 

So to put it all together a full query might look like

```rust
QUERY GetUsersThatHavePostedAndPosts => 
       GET User <- V::User(userID)
               WHERE (
                       User::Has(COUNT ?::OutE)   // OR could be User::Has(?::OutE::Count) OR User::Has(Count(?::OutE))
               )
               Posts <- User::Out::Created
       RETURN User, Posts
```
Would return `[{ User: {...}, Posts: [{...},...]}, [User: {...}, Posts: [{...}, ...] }]`

**Reply by willbowden** [reply]
Does `User::Has(COUNT ?::OutE)` mean where user has >0 out edges?

**Reply by xav-db** [reply]
Yes

**Reply by willbowden** [reply]
I think that's too hard to understand from the syntax of the language alone. Surely there's no benefit to omitting a boolean comparison like `COUNT User::InE > 0`

In fact I think most of the (all caps) query keywords should be on the outer most reasonable level of a query statement, to help with readability.

**Reply by xav-db** [reply]
> I think that's too hard to understand from the syntax of the language alone. Surely there's no benefit to omitting a boolean comparison like `COUNT User::InE > 0`
> 
> In fact I think most of the (all caps) query keywords should be on the outer most reasonable level of a query statement, to help with readability.

Fair point. We could do something like
```rust
QUERY GetUsersThatHavePostedAndPosts(userID: String) => 
       GET User <- V::User(userID)
               WHERE (
                       User::Has(?::OutE::Count()::Gt(0))  // or User::Has(?::OutE::Count() >0)
               )
               Posts <- User::Out::Created
       RETURN User, Posts
```

**Reply by willbowden** [reply]
Yeah I like that, and I think we can definitely keep boolean operators in.

If we're using a `Has` check on the `User:;` type, is `?` even necessary in the predicate? Surely the fact we're starting from the user is inferred by the fact it's `User::Has()`?

**Reply by xav-db** [reply]
Yes the user node as the source for the inner traversal is inferred but the `?` is to signify an anonymous traversal i.e a traversal with no side effects. in gremlin this would be 
`g.V(userID).has(__.outE().count().....`. 


At the end of this traversal 
> `User::Has(OutE::Count()::Gt(0))` 

you would be on the outgoing edges of the user (assuming the OutE traversed from the User).
Meaning you wouldn't be able to directly do 
> `Posts <- User::Out::Created` 

On the next line.

**Reply by willbowden** [reply]
Ahh okay I understand the need for anonymous traversal now.

Looks good to me!

**Reply by Gand4lf** [reply]
You left out the parameter...
e.g:
QUERY GetFollowers(userID: String)

**Comment by Gand4lf** [comment]
Schema example. Schema would have to be in file called `schema.hx`

```rust
N::User {
   Name: String,
   Age: Integer,
}

E::Follows{
   From: User,
   To: User,
   Properties{
   }
}
```

**Comment by swaploard** [comment]
@xav-db Open some issues for contributors.
